### DISCUSSION TITLE
How to migrate this call site of `Completion.Completer` from 3.4.2 to 3.5.0?

### DISCUSSION THREAD

**Discussion by lihaoyi** [scala/scala3#21579] [Compiler internals Q&A]
This is coming up in the Ammonite Scala 3.5.0 upgrade https://github.com/com-lihaoyi/Ammonite/pull/1548:

```scala
class DeepCompleter(
    mode: Completion.Mode,
    prefix: String,
    pos: SourcePosition
  ) extends Completion.Completer(mode, prefix, pos)
```

The new signature seems totally different: `prefix` is gone, and it takes a `matches` callback.

How what does the new signature mean, and how do I adapt the old use site to fit the new signature?

Thanks!

**Comment by lihaoyi** [comment]
Setting `matches` to `_.startsWith(prefix)` seems to make my tests pass, but is that the right thing to do? And what's this new `untpdPath: List[untpd.Tree]` parameter meant to be?

**Reply by tgodzik** [reply]
I think these changes were made so that we can do a fuzzy match instead of just prefix one. Doing `_.startsWith` should be equivalent to the previous behaviour.

The untpdPath might be needed for cases with extensions where we want to get the parsed tress instead of the desugared ones

@rochala can probably add more

**Reply by rochala** [reply]
The new fields were added as necessary optimizations for presentation compiler, so we don't recompute those fields. I've left the original method in `Completion.scala` so feel free to use it if you don't have fields precomputed: https://github.com/scala/scala3/blob/48918580e887664c857de551e7bc561fff12303e/compiler/src/dotty/tools/dotc/interactive/Completion.scala#L55

**Reply by rochala** [reply]
The API of the interactive package in dotty will evolve over time. It is still internal API. I'm trying to keep the backward compatibility, but I recommend that you migrate to the actual presentation compiler in the future. You may not need all the features, but it is way richer with features such as case completions, Scala-CLI / ammonite directives completions etc. It is also very easy to plug in. If you're interested I can help you with that / or check Scastie as it also relies on the presentation compiler. It is also MiMa tested so you shouldn't have any issues in the near future.

**Reply by rochala** [reply]
And if you want to use the `Completion.Completer` `untpdPath` is computed by running: https://github.com/rochala/dotty/blob/0b45f44092d9c00f2e3f89d08eabccb42d78ad67/presentation-compiler/src/main/dotty/tools/pc/completions/CompletionProvider.scala#L61 and `matches` is a predicate used to filter completions. You can provide your own fuzzy matcher or just do `_.startsWith` as Tomasz suggested
